### ISSUE TITLE
XVLA folding checkpoint camera setup

### ISSUE THREAD

**Issue by brycegoh** [huggingface/lerobot#2846]
### Ticket Type

❓ Technical Question

Hi, I noticed that the [lerobot/xvla-folding](https://huggingface.co/lerobot/xvla-folding) checkpoint enforced 1 empty camera and also does not label which camera is the top, left wrist or right wrist cameras as shown in the folding dataset [lerobot/xvla-soft-fold](https://huggingface.co/datasets/lerobot/xvla-soft-fold)

Referencing [config.json](https://huggingface.co/lerobot/xvla-folding/blob/main/config.json)
```json
{
"input_features":{
   "observation.images.image":{ # is this supposed to be mapped to top cam?
      "type":"VISUAL",
      "shape":[
         3,
         256,
         256
      ]
   },
   "observation.images.image2":{ # is this supposed to be mapped to one of the wrist cam?
      "type":"VISUAL",
      "shape":[
         3,
         256,
         256
      ]
   },
   "observation.images.empty_camera_0":{ # is this supposed to be mapped to one of the wrist cam?
      "type":"VISUAL",
      "shape":[
         3,
         224,
         224
      ]
   }
},
    "num_image_views": 3,
    "empty_cameras": 1, # enforces 1 empty camera, this should be changed if there is actually 3 cameras?
}
```

Given that I want to finetune from the folding checkpoint but with 3 cameras, should I change the `empty_cameras` to 0? Also, wouldn't it be better if we knew which image keys (image, image2, empty_camera_0) mapped to either top, left wrist or right wrist?

**Comment by Saad-Mallebhari** [comment]
@brycegoh 
Great question ,this is a very reasonable point, and the confusion is understandable given how the XVLA config is currently structured.

Yes, if you are fine-tuning with three real cameras, you should set `empty_cameras = 0`**, but there are a few important caveats to be aware of.

In the `xvla-folding` checkpoint, the model was trained with a *fixed number of visual inputs*. Even though only two physical cameras were used in the dataset, the architecture expects three image tensors, which is why `empty_camera_0` exists. That entry is effectively a placeholder used to keep the input shape consistent with the model’s design — it does not correspond to a real camera.

If you are fine-tuning with three actual cameras, then:
- `empty_camera_0` should be removed or replaced
- `num_image_views` should remain `3`
- all three image inputs should be populated with real data

That said, one current limitation is that the config does not explicitly document which image key corresponds to which camera (e.g., top vs wrist). In practice, this mapping is implicit and tied to how the original dataset was constructed, which makes fine-tuning on custom setups harder than it should be.

It would definitely help if the checkpoint or documentation clarified:
- whether `image`, `image2`, and `empty_camera_0` have fixed semantic meaning
- which camera viewpoints they correspond to in the folding dataset
- whether reordering or replacing them is officially supported

For now, the safest approach when fine-tuning is to:
1. Keep the number of image inputs consistent with the checkpoint
2. Explicitly control the camera-to-key mapping in your dataset
3. Avoid leaving an `empty_camera_*` entry unless you truly intend it to be unused

**Comment by jadechoghari** [comment]
Hey! @Saad-Mallebhari summarized the empty camera behavior perfectly. We’re planning to improve this soon, along with clearer documentation explaining how to work with empty vs. full camera setups.

In the meantime, here’s how cameras are mapped for the X-VLA checkpoints (this applies to all of them on the Hub):
Index | API key | Client name | Physical camera (ROS) | Role
-- | -- | -- | -- | --
0 | image0 | cam_high | front (camera_f) | Top/base
1 | image1 | cam_left_wrist | left (camera_l) | Left wrist
2 | image2 | cam_right_wrist | right (camera_r) | Right wrist

Closing this as it's addressed, feel free to open whenever!
